flow/FastAlloc.cpp

```diff
@@ -224,25 +224,35 @@ struct FastAllocator<Size>::GlobalData {
 	long long totalMemory;
 	long long partialMagazineUnallocatedMemory;
 	long long activeThreads;
-	GlobalData() : totalMemory(0), partialMagazineUnallocatedMemory(0), activeThreads(0) { 
+	GlobalData() : totalMemory(0), partialMagazineUnallocatedMemory(0), activeThreads(0) {
 		InitializeCriticalSection(&mutex);
 	}
 };
 
 template <int Size>
 long long FastAllocator<Size>::getTotalMemory() {
-	return globalData()->totalMemory;
+	EnterCriticalSection(&globalData()->mutex);
+	long long total = globalData()->totalMemory;
```

Comment:
Why do we need a mutex here?  (Aligned 64 bit reads/writes cannot tear on x86 or arm.)  Did a sanitizer ask for this?

Anyway, std::atomic might be a better choice.

Reply:
TSAN highlighted these accesses as races. Writes to all of these global data values are already protected by this mutex, e.g.:

https://github.com/apple/foundationdb/blob/master/flow/FastAlloc.cpp#L428
https://github.com/apple/foundationdb/blob/master/flow/FastAlloc.cpp#L414
https://github.com/apple/foundationdb/blob/master/flow/FastAlloc.cpp#L428

Whether that mutex should be changed to be atomics is outside the scope of this fix. My initial attempt to rip this fix out of this PR failed, but I would still ask that discussion of the thread safety work be moved to https://github.com/apple/foundationdb/pull/3523.

Reply:
For simple reads (not sure about the potentially-torn arithmetic involving `size()`), the TSAN errors are false positives on x86 and arm, at least.  In particular, at mutex release, the compiler spills registers for potentially shared state, and the processor flushes the write buffer (speaking in x86 terms), so the write is guaranteed to be globally visible once the mutex is released.  If the value is read while the mutex is held, readers will get the old value or the new value depending on whether the write was flushed for other reasons.  ARM's semantics are a bit weaker, but shouldn't cause any trouble here.

Grabbing the mutexes incurs a bit of a performance hit, but (worse) unnecessary synchronization logic is a maintenance nightmare (Why is this useless-looking mutex here?  Can I delete it?)  It's probably better to use `MOZ_TSAN_BLACKLIST`(and add a comment) to tell TSAN these functions are OK.

It's known to be impossible to implement a sound and complete race detection algorithm, so it's not surprising TSAN is giving false positives.  (Race detection is an ill-posed problem in the first place; I could go on for hours...)